station-dispenser-api-v0/src/main/java/net/modificationstation/stationapi/mixin/dispenser/block/DispenserBlockMixin.java

```diff
@@ -0,0 +1,62 @@
+package net.modificationstation.stationapi.mixin.dispenser.block;
+
+import net.minecraft.block.DispenserBlock;
+import net.minecraft.block.entity.BlockEntity;
+import net.minecraft.block.entity.DispenserBlockEntity;
+import net.minecraft.item.ItemStack;
+import net.minecraft.world.World;
+import net.modificationstation.stationapi.api.StationAPI;
+import net.modificationstation.stationapi.api.item.CustomDispenseBehavior;
+import net.modificationstation.stationapi.api.item.DispenseEvent;
+import net.modificationstation.stationapi.api.item.DispenseUtil;
+import net.modificationstation.stationapi.impl.dispenser.DispenserInfoStorage;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Unique;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.Redirect;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import java.util.Random;
+
+@Mixin(DispenserBlock.class)
+public class DispenserBlockMixin {
+    @Unique
+    DispenserBlockEntity dispenserBlockEntity;
+    @Unique
+    ItemStack currentItemStack;
+
+    @Inject(method = "method_1774", at = @At("HEAD"), cancellable = true)
+    private void dispense(World world, int x, int y, int z, Random random, CallbackInfo ci) {
+        dispenserBlockEntity = (DispenserBlockEntity) world.method_1777(x, y, z);
+        currentItemStack = dispenserBlockEntity.method_665();
+        int slot = DispenserInfoStorage.slot;
+        ItemStack[] inventory = DispenserInfoStorage.inventory;
+
+        DispenseUtil util = new DispenseUtil(world, currentItemStack, dispenserBlockEntity, inventory, slot);
+
+        if (StationAPI.EVENT_BUS.post(
+                DispenseEvent.builder().dispenseUtil(util).build()
+        ).isCanceled()) {
+            ci.cancel();
+        }
+        else {
+            if (currentItemStack != null) {
+                if (currentItemStack.getItem() instanceof CustomDispenseBehavior behavior) {
+                    behavior.dispense(util);
+                    ci.cancel();
+                }
+            }
+        }
```

Comment:
This else closure should be in its own listener instead

Reply:
ok good idea, will do